transport/src/main/java/io/netty/channel/EventLoopGroups.java

```diff
@@ -0,0 +1,179 @@
+                return constructor.newInstance(nThreads, threadFactory);
+            } catch (Exception e) {
+                logger.debug("Cannot invoke event loop group constructor: {}", constructor, e);
+            }
+        }
+        throw new IllegalStateException(
+                "Cannot successfully invoke any of discovered createEventLoopGroupConstructors: " +
+                        EVENT_LOOP_CONSTRUCTORS);
+    }
+
+    /**
+     * Tries to load {@link EventLoopGroup} class.
+     *
+     * @param fqcn event loop group fully qualified class name.
+     * @return event loop group class on success, otherwise <code>null</code>.
+     * @see #loadClass(String, Class)
+     */
+    protected static Class<? extends EventLoopGroup> loadEventLoopGroupClass(String fqcn) {
+        return loadClass(fqcn, EventLoopGroup.class);
+    }
+
+    /**
+     * Tries to load class with specified fully qualified class name.
+     *
+     * @param fqcn      fully qualified class name.
+     * @param classType class into which loaded class should be cast.
+     * @param <T>       class type.
+     * @return class on success, otherwise <code>null</code>.
+     */
+    protected static <T> Class<T> loadClass(String fqcn, Class<T> classType) {
+        try {
+            Class<?> clazz = Class.forName(fqcn);
+            return (Class<T>) clazz.asSubclass(classType);
+        } catch (ClassNotFoundException e) {
+            logger.trace("Unable to load class {}: class is not present on classpath.", fqcn);
+            return null;
+        }
+    }
+
+    protected static String ucFirst(String s) {
+        char[] chars = s.toCharArray();
+        chars[0] = Character.toUpperCase(chars[0]);
+        return new String(chars);
+    }
+
+    private static List<Constructor<? extends EventLoopGroup>> createEventLoopGroupConstructors() {
+        List<Constructor<? extends EventLoopGroup>> res = new ArrayList<Constructor<? extends EventLoopGroup>>(4);
+        createEventLoopGroupConstructors(nativeImplementations(), res);
+        createEventLoopGroupConstructors(standardImplementations(), res);
+        return Collections.unmodifiableList(res);
+    }
+
+    private static void createEventLoopGroupConstructors(List<String> implementations,
+                                                         List<Constructor<? extends EventLoopGroup>> dst) {
+        for (String impl : implementations) {
+            try {
+                Class<? extends EventLoopGroup> clazz = loadEventLoopGroupClass(eventLoopGroupClass(impl));
+                Constructor<? extends EventLoopGroup> constructor =
+                        clazz.getConstructor(int.class, ThreadFactory.class);
+                dst.add(constructor);
+            } catch (Exception e) {
+                // no one cares
+            }
+        }
+    }
+
+    protected static String eventLoopGroupClass(String implementation) {
+        String pkg = EventLoopGroups.CHANNEL_PACKAGE + "." + implementation;
```

Comment:
Manually building strings which we assume will map to classes that can be loaded feels fragile. There may be shading or other manipulation of classes that is difficult to account for at compile time. IMO we should use the source of truth (type information provided by the runtime e.g. `Class.getName()`) unless there is good reason not to, or it is otherwise unavoidable.